Civi/Core/LogManager.php

```diff
@@ -39,6 +39,9 @@ public function getLog($channel = 'default') {
       $c = \Civi::container();
       $svc = "log." . $channel;
       $this->channels[$channel] = $c->has($svc) ? $c->get($svc) : $c->get(self::DEFAULT_LOGGER);
+      if ($channel !== 'default' && method_exists($this->channels[$channel], 'setChannel')) {
+        $this->channels[$channel]->setChannel($channel);
+      }
```

Comment:
@eileenmcnaughton - OK, I'm personally happy with a policy of splitting channels into separate files by default.

From an `r-tech` POV, I suspect that some sysadmins/upgraders might prefer an option to keep the unified log (e.g. if they've already linked that into their log-aggregation system). Would be curious if we had some feedback from folks with sophisticated deployments (e.g. @MegaphoneJon @adixon @mlutfy).

Maybe a setting where you could opt-out/opt-in to split-files. Ex:

```php
  // Default - Log every channel to separate file: "log_file_channels" == "/.*/"
  // Alternative - Don't log anything to a separate file: "log_file_channels" == "/^$/"
  // Alternative - Log two specific channels to separate files: "log_file_channels" == "/^(mail|ipn)$/

  if (preg_match(Civi::settings()->get('log_file_channels'), $channel)) {
    // Use separate file
  }
  else {
    // Use traditional file
  }
```

Implementation-wise, there should be a separate instance of the logger for each channel. To my eye, this looks like it'll re-use one instance, which could cause some crossed wires. Consider:

```php
class Foo {
  private $log;
  public function __construct() { $this->log = Civi::log('foo'); }
  public function fooify() { $this->info("Hello from Foo"); }
}

class Bar {
  private $log;
  public function __construct() { $this->log = Civi::log('bar'); }
  public function barbecueTofu() { $this->info("Hello from Bar"); }
}

$foo = new Foo();
$bar = new Bar();
$foo->fooify();
$bar->barbecueTofu();
// If the two objects reference the same `$log`, then one or the other will route to the wrong place.
``` 

Assuming you're keen on the combination of `DEFAULT_LOGGER`  and `method_exists(...'setChannel')` -- then one could get the desired result with `clone`, e.g.

```php
if ($c->has($svc) {
  $log = $c->get($svc);
}
else {
  $log =  $c->get(self::DEFAULT_LOGGER);;
  if ($channel !== 'default' && method_exists($this->channels[$channel], 'setChannel') /* && settings encourage split files */) {
    $log = clone $log;
    $log->setChannel($channel);
  }
}
$this->channels[$channel] = $log;
```


(Also, small note - the old code there was using a default value of empty-string (`''`). The property `protected $channel` has a default of `NULL`. Dunno if that makes a difference here.)

Reply:
It's worth noting that there are no existing configured channels so far - so default behaviour is no change - however there would be a change if we also defined channels

If this needs a setting to maintain then I don't think we should merge that change - it makes more sense to manage it all in an extension - which I'm making good progress on.

The main reason I thought it might be worth merging is the splitting out of log files give some incentive for people to think about what should get logged to which channels

Based on your feedback I'm inclined to close this but I'll leave it for a couple of days just to let people add any thoughts

Reply:
Proposed variant: #20163